src/v/cluster/archival_metadata_stm.cc

```diff
@@ -851,6 +852,42 @@ archival_metadata_stm::get_segments_to_cleanup() const {
     using lw_segment_meta = cloud_storage::partition_manifest::lw_segment_meta;
     std::vector<lw_segment_meta> backlog = _manifest->lw_replaced_segments();
 
+    // Make sure that 'replaced' list doesn't have any references to active
+    // segments. This is a protection from the data loss.
+    auto backlog_size = backlog.size();
+    backlog.erase(
+      std::remove_if(
+        backlog.begin(),
+        backlog.end(),
+        [this](const lw_segment_meta& m) {
+            auto it = _manifest->find(m.base_offset);
+            if (it == _manifest->end()) {
+                return false;
+            }
+            const auto& s = it->second;
+            if (m.size_bytes == 0) {
+                // Legacy name format, these segments are always reuploaded
```

Comment:
This and the following check seem a bit fragile: they rely on knowledge of how names are composed and could get out of date as we evolve the format.

What if instead, we just generate the key for `s` and for `m`, and string compare them?  That seems like it would be a pretty bulletproof version, and IMO a worthwhile efficiency cost, as we'll only hit this path  when replacing segments with the same base offset.

Reply:
did that